File: Sources/Sentry/SentryReplayRecording.m
#import "SentryReplayRecording.h"
#import "SentryDateUtil.h"

NS_ASSUME_NONNULL_BEGIN

@implementation SentryReplayRecording

- (instancetype)initWithSegmentId:(NSInteger)segmentId
                             size:(NSInteger)size
                            start:(NSDate *)start
                         duration:(NSTimeInterval)duration
                       frameCount:(NSInteger)frameCount
                        frameRate:(NSInteger)frameRate
                           height:(NSInteger)height
                            width:(NSInteger)width
{
    if (self = [super init]) {
        self.segmentId = segmentId;
        self.size = size;
        self.start = start;
        self.duration = duration;
        self.frameCount = frameCount;
        self.frameRate = frameRate;
        self.height = height;
        self.width = width;
    }
    return self;
}

- (NSDictionary *)headerForReplayRecording
{
    return @{ @"segment_id" : @(self.segmentId) };
}

- (NSArray<NSDictionary<NSString *, id> *> *)serialize
{

    long timestamp = [SentryDateUtil millisecondsSince1970:self.start];

    // This format is defined by RRWeb
    // empty values are required by the format
    NSDictionary *metaInfo = @{
        @"type" : @4,
        @"timestamp" : @(timestamp),
        @"data" : @ { @"href" : @"", @"height" : @(self.height), @"width" : @(self.width) }
    };

    NSDictionary *recordingInfo = @{
        @"type" : @5,
        @"timestamp" : @(timestamp),
        @"data" : @ {
            @"tag" : @"video",
            @"payload" : @ {
                @"segmentId" : @(self.segmentId),
                @"size" : @(self.size),
                @"duration" : @(self.duration),
                @"encoding" : SentryReplayEncoding,
                @"container" : SentryReplayContainer,
                @"height" : @(self.height),
                @"width" : @(self.width),
                @"frameCount" : @(self.frameCount),
                @"frameRateType" : SentryReplayFrameRateType,
                @"frameRate" : @(self.frameRate),
                @"left" : @0,
                @"top" : @0,
            }
        }
    };

    return @[ metaInfo, recordingInfo ];
}

@end

NS_ASSUME_NONNULL_END

File: Sources/Swift/Integrations/SessionReplay/SentryReplayRecording.swift
@_implementationOnly import _SentryPrivate
import Foundation

@objcMembers
class SentryReplayRecording: NSObject {
    
    static let SentryReplayEncoding = "h264"
    static let SentryReplayContainer = "mp4"
    static let SentryReplayFrameRateType = "constant"
    
    let segmentId: Int

    let meta: SentryRRWebMetaEvent
    let video: SentryRRWebVideoEvent
    
    init(segmentId: Int, size: Int, start: Date, duration: TimeInterval, frameCount: Int, frameRate: Int, height: Int, width: Int) {
        self.segmentId = segmentId
        
        meta = SentryRRWebMetaEvent(timestamp: start, height: height, width: width)
        video = SentryRRWebVideoEvent(timestamp: start, segmentId: segmentId, size: size, duration: duration, encoding: SentryReplayRecording.SentryReplayEncoding, container: SentryReplayRecording.SentryReplayContainer, height: height, width: width, frameCount: frameCount, frameRateType: SentryReplayRecording.SentryReplayFrameRateType, frameRate: frameRate, left: 0, top: 0)
    }

    func headerForReplayRecording() -> [String: Any] {
        return ["segment_id": segmentId]
    }

    func serialize() -> [[String: Any]] {
        let metaInfo = meta.serialize()

        let recordingInfo = video.serialize()

        return [metaInfo, recordingInfo]
    }
}
